### DISCUSSION TITLE
Query Konwledge base over API

### DISCUSSION THREAD

**Discussion by WhyNotPal** [mitre/caldera#2024] [Q&A] [answered]
Dear, 
Thanks for sharing Caldera with us, its potential is amazing. 
I had a quick look around and was interested in the cli operations. However, if I got this right, it's possible to operate at agent/adversaries/operation level, but it's not possible to query the knowledge base - so no way to get the knowledge acquired via the API. Is that right? 

Do you have any plan in that respect? 
Thanks

**Comment by uruwhy** [comment, marked as answer]
Hello,

We're glad you are enjoying our tool! Just so I'm on the same page, are you referring to facts obtained from running operations? If so, you can use the REST API to fetch results for an individual operation link (requires link ID) and also to fetch the operation report for a specified operation ID. The operation report is a large JSON that contains all the operation details, including the links that were run and any facts that were generated. The link results should contain facts for an individual link. If either of these is what you're looking for, I can look into the details of the required POST requests.

We are also in the process of pushing out a new version of our REST API, which will make it easier to query the API and find what you're looking for in terms of operations and facts.

**Reply by WhyNotPal** [reply]
Hey 
Thanks for your reply. 
I have a question in that regard. - According to the docs (https://caldera.readthedocs.io/en/latest/The-REST-API.html) the docs for the new version of the API should be available online, and via the caldera navigation menu, but I can't find it anywhere? 

Also, is there a way to delete facts sources? (Couldn't find it in the docs) - What if you want to get rid of historical facts? Unless I'm mistaken, it's not possible from the web interface nor via the API. 
Thanks.

**Reply by uruwhy** [reply]
For caldera v4 alpha release:

Accessing documentation - start your server up, and then access the `/api/docs` path (e.g. `http://localhost:8888/api/docs`). The link also shows up at the top of the original REST API docs page.

Deleting fact sources - in the fact sources page, there is a delete button to the right of the "add source" and "duplicate" button. It should have the trash icon. You can also do this via the API by sending a DELETE HTTP request (e.g. `curl -X DELETE -H "KEY:$YOUR_API_KEY" http://mycalderaserver:8888/api/v2/sources/$ID_OF_SOURCE_TO_DELETE`)

**Reply by WhyNotPal** [reply]
Hey 
Thanks for the quick reply! 
Thanks for the link for the api/doc- Will explore very soon. 

QQ however - the last release is 3.1.0 (If I read correctly from 
![image](https://user-images.githubusercontent.com/66071761/146400750-e714eff0-5a98-48c1-ab42-9a7d2b77dc20.png)
Is there an easy way to get V4? Thanks

**Reply by uruwhy** [reply]
Sure thing - 3.1.0 is the last stable release. We are currently going through a v4 alpha release - you can view the various v4 alpha releases at https://github.com/mitre/caldera/releases (e.g. https://github.com/mitre/caldera/tree/4.0.0-alpha.2)

**Reply by WhyNotPal** [reply]
Great! 
In any case, thanks so much for sharing caldera with the world !!

**Reply by uruwhy** [reply]
We're glad you're enjoying it! Let us know if you have any other questions
